[nnc] Fix float->bool conversion on cpu

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57798 [nnc] Fix float->bool conversion on cpu**

Our instruction sequence was just plain wrong, instead of `fcmp une %x, +0.0`
(unordered equal 0.0) we were doing `fcmp uno`, which is just an unordered check
(i.e., is either side NaN).

Differential Revision: [D28276269](https://our.internmc.facebook.com/intern/diff/D28276269/)

bertmaher committed:
[nnc] Fix float->bool conversion on cpu

Our instruction sequence was just plain wrong, instead of `fcmp une %x, +0.0`
(unordered equal 0.0) we were doing `fcmp uno`, which is just an unordered check
(i.e., is either side NaN).

Differential Revision: [D28276269](https://our.internmc.facebook.com/intern/diff/D28276269/)

[ghstack-poisoned]

navahgar commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -1087,7 +1087,7 @@ class TensorExprFuser {
     // A hook to optimizations limitter to allow bisecting the pass
     if (!JIT_OPT_ALLOWED) {
-      return false;
+      REQ(false && "jit opt limited");
```
Does this belong to this PR?

bertmaher replied:
Yeah it's not strictly related but it was useful to me while bisecting so I thought I might as well leave it.

ZolotukhinM replied:
Nit: probably could've rewritten it as `REQ(JIT_OPT_ALLOWED && "jit opt limited")`.

bertmaher replied:
Yeah good point that's cleaner.  I'll change to that!

ZolotukhinM replied:
I'm actually not sure the macros magic is composable, but would be cool if it works :)

navahgar reviewed (approved):
Good catch. LGTM

eellison reviewed (commented):
Oop, my bad! Do you know why this wasnt caught by the test? https://github.com/pytorch/pytorch/pull/48976/files#diff-5b331823d79fc13c5436ad744a4856c5ef0acd9403ca3fa9d4a12f8ec69cdc8aR1281 I have float->bool tests there..

bertmaher commented:
> Oop, my bad! Do you know why this wasnt caught by the test? https://github.com/pytorch/pytorch/pull/48976/files#diff-5b331823d79fc13c5436ad744a4856c5ef0acd9403ca3fa9d4a12f8ec69cdc8aR1281 I have float->bool tests there..

Dang, this is a weakness in my `data_for` function that we might want to systematically remedy.  For `float` dtypes it generates data with `rand`, but this conversion only fails with a precise value of 0.0, which is unlikely as a result of `rand`.

bertmaher committed:
REQ(JIT_OPT_ALLOWED) on "[nnc] Fix float->bool conversion on cpu"


Our instruction sequence was just plain wrong, instead of `fcmp une %x, +0.0`
(unordered equal 0.0) we were doing `fcmp uno`, which is just an unordered check
(i.e., is either side NaN).

Differential Revision: [D28276269](https://our.internmc.facebook.com/intern/diff/D28276269/)

[ghstack-poisoned]
